pkg/chunk/repair_test.go

```diff
@@ -0,0 +1,296 @@
+		msg, err := trojan.NewMessage(testTopic, maxPayload)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// invoke the chunk repair handler
+		repairHandler(msg)
+
+		if pushServiceCalled {
+			t.Fatal("push service called even if the chunk is not present")
+		}
+	})
+
+	t.Run("repair-chunk-closest-peer-not-present", func(t *testing.T) {
+		// generate test chunk, store and publisher
+		c3 := chunktesting.GenerateTestRandomChunk()
+
+		// create a mock storer
+		mockStorer := storemock.NewStorer()
+		defer mockStorer.Close()
+		_, err := mockStorer.Put(context.Background(), storage.ModePutRequest, c3)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// create a mock pushsync service
+		var receiptError error
+		pushSyncService := pushsyncmock.New(func(ctx context.Context, chunk swarm.Chunk) (*pushsync.Receipt, error) {
+			receiptError = errors.New("invalid receipt")
+			return nil, receiptError
+		})
+
+		// create the chunk repair handler
+		repairHandler := chunk.NewRepairHandler(mockStorer, logger, pushSyncService)
+
+		//create a trojan message to trigger the repair of the chunk
+		testTopic := trojan.NewTopic("foo")
+		maxPayload := make([]byte, swarm.SectionSize)
+		var msg trojan.Message
+		copy(maxPayload, c3.Address().Bytes())
+		msg, err = trojan.NewMessage(testTopic, maxPayload)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// invoke the chunk repair handler
+		repairHandler(msg)
+
+		if receiptError == nil {
+			t.Fatal("pushsync did not generate a receipt error")
+		}
+	})
+}
+
+// newTestNetStore creates a test store with a set RemoteGet func
+func newTestNetStore(t *testing.T, recoveryFunc chunk.RecoveryHook) storage.Storer {
+	t.Helper()
+	storer := mock.NewStorer()
+	logger := logging.New(ioutil.Discard, 5)
+
+	mockStorer := storemock.NewStorer()
+	serverMockAccounting := accountingmock.NewAccounting()
+	price := uint64(10)
+	pricerMock := accountingmock.NewPricer(price, price)
+	peerID := swarm.MustParseHexAddress("9ee7add7")
+	ps := mockPeerSuggester{eachPeerRevFunc: func(f topology.EachPeerFunc) error {
+		_, _, _ = f(peerID, 0)
+		return nil
+	}}
+	server := retrieval.New(retrieval.Options{
+		Storer:     mockStorer,
+		Logger:     logger,
+		Accounting: serverMockAccounting,
+		Pricer:     pricerMock,
+	})
+	recorder := streamtest.New(
+		streamtest.WithProtocols(server.Protocol()),
+	)
+	retrieve := retrieval.New(retrieval.Options{
+		Streamer:    recorder,
+		ChunkPeerer: ps,
+		Storer:      mockStorer,
+		Logger:      logger,
+		Accounting:  serverMockAccounting,
+		Pricer:      pricerMock,
+	})
+
+	netStore := netstore.New(storer, recoveryFunc, nil, retrieve, logger, nil)
+	return netStore
+}
+
+type mockPeerSuggester struct {
+	eachPeerRevFunc func(f topology.EachPeerFunc) error
+}
+
+func (s mockPeerSuggester) EachPeer(topology.EachPeerFunc) error {
+	return errors.New("not implemented")
+}
+func (s mockPeerSuggester) EachPeerRev(f topology.EachPeerFunc) error {
+	return s.eachPeerRevFunc(f)
+}
```

Comment:
can't we import these from `retrieval`?

Reply:
not from retrieval_test

Reply:
maybe i am missing something here, is there a problem with exporting these from the test package rather than duplicating the code?

Reply:
It is problem exporting with test packages